### PR TITLE
feat(asm): 1-click activation with RCM

### PR DESCRIPTION
**Pull request by avara1986**
## Description
Remote Config Management: 1-click activation

## Reviewer Checklist
- [ ] Title is accurate.
- [ ] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] PR cannot be broken up into smaller PRs.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: ddtrace/settings/config.py [hunk 1]
```diff
@@ -223,7 +224,7 @@ def __init__(self):
         self._trace_compute_stats = asbool(
             os.getenv("DD_TRACE_COMPUTE_STATS", os.getenv("DD_TRACE_STATS_COMPUTATION_ENABLED", False))
         )
-        self._appsec_enabled = asbool(os.getenv("DD_APPSEC_ENABLED", False))
+        self._appsec_enabled = asbool(os.getenv(APPSEC_ENV, False))
```
Comment (P403n1x87) [inline ddtrace/settings/config.py:227]: maybe worth using `envier` now that it is a dependency?

File: ddtrace/appsec/remoteconfiguration.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+def _rc_features_is_enabled():
+    return asbool(os.environ.get(RC_ENV)) and (
+        APPSEC_ENV not in os.environ or asbool(os.environ.get(APPSEC_ENV)) is True
+    )
+
+
+def appsec_enable_rc(tracer):
+    if _rc_features_is_enabled():
+        if not hasattr(tracer, "_rc"):
+            tracer._rc = RemoteConfig()
```
Comment (P403n1x87) [inline ddtrace/appsec/remoteconfiguration.py:23]: This already implements a singleton pattern, so registering the product should be enough

File: ddtrace/appsec/remoteconfiguration.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        tracer._rc.register(ASM_FEATURES_PRODUCT, appsec_rc_reload_features(tracer))
```
Comment (P403n1x87) [inline ddtrace/appsec/remoteconfiguration.py:24]:
```suggestion
        RemoteConfig.register(ASM_FEATURES_PRODUCT, appsec_rc_reload_features(tracer))
```

File: ddtrace/appsec/remoteconfiguration.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import os
+
+from ddtrace.constants import APPSEC_ENV
+from ddtrace.constants import RC_ENV
```
Comment (P403n1x87) [inline ddtrace/appsec/remoteconfiguration.py:4]: I think the plan is to remove this as it would break those products that rely on RCM.

File: ddtrace/internal/remoteconfig/client.py [hunk 5]
```diff
@@ -127,6 +129,12 @@ class AgentPayload(object):
     client_configs = attr.ib(type=Set[str], default={})
 
 
+def _load_json(data):
+    if (3, 6) > sys.version_info > (3,) and isinstance(data, six.binary_type):
```
Comment (P403n1x87) [inline ddtrace/internal/remoteconfig/client.py:133]: can we add type annotations?

File: ddtrace/internal/remoteconfig/client.py [hunk 6]
```diff
@@ -127,6 +129,12 @@ class AgentPayload(object):
     client_configs = attr.ib(type=Set[str], default={})
 
 
+def _load_json(data):
```
Comment (P403n1x87) [inline ddtrace/internal/remoteconfig/client.py:132]: can we add type annotations?

**Comment by juanjux** [comment]
Great job tackling this beast. Nothing to add outside of the existing comments.

File: ddtrace/appsec/remoteconfiguration.py (hunk 4 above)
Comment (avara1986) [reply ddtrace/appsec/remoteconfiguration.py:4]: remove the constants? or what do you mean? TY!

File: ddtrace/settings/config.py (hunk 1 above)
Comment (avara1986) [reply ddtrace/settings/config.py:227]: It's interesting `envier` :clap: I think this refactor it could be better in other PR, what do you think?

File: ddtrace/appsec/remoteconfiguration.py (hunk 4 above)
Comment (P403n1x87) [reply ddtrace/appsec/remoteconfiguration.py:4]: Sorry, I meant the environment variable. From what I understood from RCM, there is no plan for exposing such a variable, as setting it to false would break all products that rely on RCM. Eventually, we would then want to get rid of this constant as it will be unused.

File: ddtrace/internal/remoteconfig/constants.py [hunk 7]
```diff
@@ -0,0 +1,3 @@
+ASM_FEATURES_PRODUCT = "FEATURES"
+# TODO(alberto.vara) [2022-09-21] in few weeks `FEATURES` will be renamed to `ASM_FEATURES`
+# ASM_FEATURES_PRODUCT = "ASM_FEATURES"
```
Comment (gnufede) [inline ddtrace/internal/remoteconfig/constants.py:3]: If this is final, why not do this already?

File: ddtrace/constants.py [hunk 8]
```diff
@@ -33,6 +34,7 @@
 APPSEC_WAF_TIMEOUTS = "_dd.appsec.waf.timeouts"
 APPSEC_WAF_VERSION = "_dd.appsec.waf.version"
 
+APPSEC_RC_ENABLED_ENV = "DD_APPSEC_RC_ENABLED"
```
Comment (P403n1x87) [inline ddtrace/constants.py:37]: 🙌

**Review by juanjux — APPROVED** [review]

File: tests/internal/test_remoteconfig.py [hunk 9]
```diff
@@ -30,3 +111,22 @@ def test_remote_config_forksafe():
...
+            self.features = features
+
+    callback = Callback()
+
+    mock_send_request.return_value = get_mock_encoded_msg(b'{"asm":{"enabled":true}}')
+    rc = RemoteConfig()
+    rc.register(ASM_FEATURES_PRODUCT, callback._reload_features)
+    sleep(3)
+    mock_send_request.assert_called_once()
+    print(callback.features)
```
Comment (gnufede) [inline tests/internal/test_remoteconfig.py:131]: Are the `sleep` and the `print` required for the test? Maybe we can at least reduce the sleep time?

**Review by gnufede — APPROVED** [review]

File: tests/internal/test_remoteconfig.py (hunk 9 above)
Comment (avara1986) [reply tests/internal/test_remoteconfig.py:131]: I tried with 1sec and RC wasn't working, 2sec random success and 3sec was the minimun value with 100% success rate

**Review by gnufede — APPROVED** [review]

File: tests/internal/test_remoteconfig.py (hunk 9 above)
Comment (P403n1x87) [reply tests/internal/test_remoteconfig.py]: The RCM polling interval can be controlled with `DD_REMOTECONFIG_POLL_SECONDS`. Perhaps setting that to e.g. 0.5 before registering might help reduce the sleep time?

Comment (avara1986) [reply tests/internal/test_remoteconfig.py]: :clap: good idea @P403n1x87

**Review by gnufede — APPROVED** [review]

**Review by juanjux — APPROVED** [review]

File: ddtrace/appsec/remoteconfiguration.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+import os
```
Comment (Kyle-Verhoog) [inline ddtrace/appsec/remoteconfiguration.py:1]: This module is in the public API now which means `enable_appsec_rc` and `appsec_rc_reload_features` are public methods. Is this intentional?

File: setup.py [hunk 11]
```diff
@@ -250,7 +250,7 @@ def get_exts_for(name):
         "protobuf>=3,<4.0; python_version=='3.6'",
         "protobuf>=3,<3.18; python_version<'3.6'",
         "tenacity>=5",
-        "attrs>=19.2.0",
+        "attrs>=21.4.0",
```
Comment (Kyle-Verhoog) [inline setup.py:253]: Why was this change necessary? Should we include an `upgrade` release note that our minimum version support for `attrs` changed?

**Review by Kyle-Verhoog — COMMENTED** [review]
Sorry for the late review! I had a couple questions

File: ddtrace/appsec/remoteconfiguration.py (hunk 10 above)
Comment (avara1986) [reply ddtrace/appsec/remoteconfiguration.py:1]: No it isn't, could i `remoteconfiguration.py` to `remoteconfiguration.py` right?

File: setup.py (hunk 11 above)
Comment (avara1986) [reply setup.py:253]: @Kyle-Verhoog CI started to fail with this error:

```
  File "/root/project/.riot/venv_py3103/lib/python3.10/site-packages/cattr/disambiguators.py", line 15, in <module>
    from cattrs._compat import get_origin
  File "/root/project/ddtrace/internal/module.py", line 190, in _exec_module
    self.loader.exec_module(module)
  File "/root/project/.riot/venv_py3103/lib/python3.10/site-packages/cattrs/_compat.py", line 16, in <module>
    from attr import resolve_types
ImportError: cannot import name 'resolve_types' from 'attr'
```

https://app.circleci.com/pipelines/github/DataDog/dd-trace-py/21452/workflows/b043126a-d4c8-49d3-84cc-703e95225486/jobs/1460754

File: ddtrace/appsec/remoteconfiguration.py (hunk 10 above)
Comment (Kyle-Verhoog) [reply ddtrace/appsec/remoteconfiguration.py:1]: Yeah I think we can make the module internal 👍

File: setup.py (hunk 11 above)
Comment (Kyle-Verhoog) [reply setup.py:253]: Is there an older version of attrs we can use (preferably the oldest)? We should be as unrestrictive as possible with our dependency versions.

Also we should still have an `upgrade` note since this could be a breaking change for users.
